scripts/ci-report/main.go

```diff
@@ -0,0 +1,214 @@
+package main
+
+import (
+	"encoding/json"
+	"errors"
+	"fmt"
+	"io"
+	"os"
+	"strings"
+	"time"
+
+	"golang.org/x/exp/slices"
+	"golang.org/x/xerrors"
+)
+
+func main() {
+	if len(os.Args) != 2 {
+		_, _ = fmt.Println("usage: ci-report <gotests.json>")
+		os.Exit(1)
+	}
+	name := os.Args[1]
+
+	goTests, err := parseGoTestJSON(name)
+	if err != nil {
+		_, _ = fmt.Printf("error parsing gotestsum report: %v", err)
+		os.Exit(1)
+	}
+
+	rep, err := parseCIReport(goTests)
+	if err != nil {
+		_, _ = fmt.Printf("error parsing ci report: %v", err)
+		os.Exit(1)
+	}
+
+	err = printCIReport(os.Stdout, rep)
+	if err != nil {
+		_, _ = fmt.Printf("error printing report: %v", err)
+		os.Exit(1)
+	}
+}
+
+func parseGoTestJSON(name string) (GotestsumReport, error) {
+	f, err := os.Open(name)
+	if err != nil {
+		return GotestsumReport{}, xerrors.Errorf("error opening gotestsum json file: %w", err)
+	}
+	defer f.Close()
+
+	dec := json.NewDecoder(f)
+	var report GotestsumReport
+	for {
+		var e GotestsumReportEntry
+		err = dec.Decode(&e)
+		if errors.Is(err, io.EOF) {
+			break
+		}
+		if err != nil {
+			return GotestsumReport{}, xerrors.Errorf("error decoding json: %w", err)
+		}
+		e.Package = strings.TrimPrefix(e.Package, "github.com/coder/coder/")
+		report = append(report, e)
+	}
+
+	return report, nil
+}
+
+func parseCIReport(report GotestsumReport) (CIReport, error) {
+	packagesSortedByName := []string{}
+	packageTimes := map[string]float64{}
+	packageFail := map[string]int{}
+	packageSkip := map[string]bool{}
+	testTimes := map[string]float64{}
+	testSkip := map[string]bool{}
+	testOutput := map[string]string{}
+	testSortedByName := []string{}
+	for i, e := range report {
+		switch e.Action {
+		// A package/test may fail or pass.
+		case Fail:
+			if e.Test == "" {
+				packageTimes[e.Package] = *e.Elapsed
+			} else {
+				packageFail[e.Package]++
+				name := e.Package + "." + e.Test
+				testTimes[name] = *e.Elapsed
+			}
+		case Pass:
+			if e.Test == "" {
+				packageTimes[e.Package] = *e.Elapsed
+			} else {
+				name := e.Package + "." + e.Test
+				delete(testOutput, name)
```

Comment:
I remember that we suspected chopped test output someday. Won't it be a problem?

> // A package/test may fail or pass

isn't that a reason why Go tests have this uknown state too?

Reply:
Ah good catch, I mixed this up with the issue having been fixed in `gotestsum`, but since we're doing our own parsing, we'll have to handle that case as well. I'll add a separate sample/golden and handle those cases.

Reply:
Fixed!